R/dependencies.R

```diff
@@ -23,3 +23,30 @@ find_dependencies <- function(expr) {
        variables = unique(variables$get()))
 
 }
+
+
+topological_order <- function(deps) {
+  if (all(lengths(deps) == 0)) {
+    return(list(success = TRUE, value = seq_along(deps)))
+  }
+
+  m <- matrix(FALSE, length(deps), length(deps))
+  for (i in seq_along(deps)) {
+    m[, i] <- unname(names(deps) %in% deps[[i]])
+  }
+
+  pending <- rep(TRUE, length(deps))
+  ret <- integer(0)
+  while (any(pending)) {
+    i <- which(pending)[colSums(m[, pending, drop = FALSE]) == 0]
```

Comment:
Here we find entries where m[1,1] = m[2,1] = m[3,1] = FALSE - meaning a does not depend on anything else. 
If we find such entries, then 'a' is known, so set m[1,1], m[1,2], m[1,3] to FALSE and don't worry about 'a' any more.

Then try again and see if we can resolve anymore - until hopefully we have no unresolved vars. 

But if we try, and can't do this to any of the vars (so `i` comes back empty) then we are cyclic and can't resolve.